### PR TITLE
Fix handling of empty request content by allowing to disable listeners per operation

### PR DESCRIPTION
**Pull request by teohhanhui**
| Q             | A
| ------------- | ---
| Bug fix?      | yes?
| New feature?  | kinda?
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1282, #2731
| License       | MIT
| Doc PR        | TODO

The toggleable listeners are:
- ReadListener ("read")
- DeserializeListener ("deserialize")
- ValidateListener ("validate")
- WriteListener ("write")
- SerializeListener ("serialize")

Basically, we fix the handling of empty request content by **removing** explicit handling of empty request content (which was wrong, resulting in weird bugs such as #2731), but allowing the user to be in control instead.

TODO:

- [x] Fix tests
- [x] Add tests
- [ ] GraphQL support

### CODE REVIEW

**Comment by soyuka** [comment]
Thanks to this you'd have the ability to remove some listeners (`@ApiResource(attributes={"read"=false})`) ?
What are the use cases? Why isn't it possible to send an empty body anymore?

**Comment by teohhanhui** [comment]
> Thanks to this you'd have the ability to remove some listeners (@ApiResource(attributes={"read"=false})) ?

Yes.

> Why isn't it possible to send an empty body anymore?

I don't get what you mean. It's possible to send an empty body if you set `"deserialize"=false`

**Comment by teohhanhui** [comment]
Well, we fix #1282 by doing the right thing out of the box, and allowing the user to disable specific listeners so that they're able to send empty body, as necessary.

**Comment by teohhanhui** [comment]
But I'll add tests to show that.

**Comment by soyuka** [comment]
> I don't get what you mean. It's possible to send an empty body if you set "deserialize"=false (and possibly "validate"=false, depending on the use case).

You removed a functional test, in a way a promise that this use case works :p. 

> Or in other words, no weird attempts by us to try to read the user's mind. see_no_evil

Couldn't we fix this by implementing the following table instead?

| content-type | body content | result | notes |
|-----------------|-----------------|---------|--------|
| not set         |  empty          |  process the request | |
| not set          | with data      | 406 | throw not acceptable: content-type is mandatory |
| set               |  empty          | may throw 400 if deserializer fails | an empty string is valid text but invalid json |
| set               | with data       | process the request ||

**Comment by teohhanhui** [comment]
> You removed a functional test, in a way a promise that this use case works :p.

Because it's wrong.

> Couldn't we fix this by implementing the following table instead?

No. The interpretation depends on the content-type. We don't have enough information to make that decision. It's better to leave it up to the user. But we do the right thing by default, by leaving that up to the Serializer.

File: features/main/crud.feature [hunk 1]
```diff
@@ -523,46 +523,6 @@ Feature: Create-Retrieve-Update-Delete
     }
     """
 
-  Scenario: Update a resource with empty body
```
Comment (dunglas) [inline features/main/crud.feature]: It's a BC break

File: src/Bridge/Symfony/Bundle/Resources/config/api.xml [hunk 2]
```diff
@@ -154,6 +154,7 @@
             <argument type="service" id="api_platform.subresource_data_provider" />
             <argument type="service" id="api_platform.serializer.context_builder" />
             <argument type="service" id="api_platform.identifier.converter" />
+            <argument type="service" id="api_platform.metadata.resource.metadata_factory" on-invalid="null" />
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Resources/config/api.xml]: `on-invalid="null"` is useless, isn't it?

File: src/Metadata/Resource/ToggleableOperationAttributeTrait.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Metadata\Resource;
+
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+
+trait ToggleableOperationAttributeTrait
```
Comment (dunglas) [inline src/Metadata/Resource/ToggleableOperationAttributeTrait.php]:
```suggestion
/**
 * @internal
 */
trait ToggleableOperationAttributeTrait
```

**Comment by dunglas** [comment]
Also, couldn't this replace the `_api_receive`, `_api_respond` and `_api_persist` request attributes?

File: features/main/crud.feature (hunk 1 above)
Comment (teohhanhui) [reply features/main/crud.feature]: Yes, but it's a bug fix. The previous behavior was wrong. We can't fix this bug without... fixing it.

File: src/Bridge/Symfony/Bundle/Resources/config/api.xml (hunk 2 above)
Comment (teohhanhui) [reply src/Bridge/Symfony/Bundle/Resources/config/api.xml]: Hmm... Indeed, `api_platform.metadata.resource.metadata_factory` must exist.

**Comment by teohhanhui** [comment]
> Also, couldn't this replace the _api_receive, _api_respond and _api_persist request attributes?

Yes, but we'd need to add the same for:

- AddFormatListener ("negotiate")
- RespondListener ("send")

WDYT @dunglas?

**Comment by teohhanhui** [comment]
The semantics would be of interest to the built-in events implementation. /cc @alanpoulain

**Comment by teohhanhui** [comment]
> Also, couldn't this replace the _api_receive, _api_respond and _api_persist request attributes?

@dunglas Another difficulty is that `_api_respond` is also used in `SwaggerUiListener` and `ExceptionListener`. I'm not sure what we should do about that.

**Comment by teohhanhui** [comment]
Could someone help implement the GraphQL support?

@alanpoulain? :smile:

**Comment by teohhanhui** [comment]
@dunglas Could GraphQL support be added in a separate PR (in `master`)? My primary aim here is to fix a bug. It's only as a result of that that we end up introducing a new feature. :laughing:

File: features/main/crud.feature [hunk 4]
```diff
@@ -91,6 +91,12 @@ Feature: Create-Retrieve-Update-Delete
     }
     """
 
+  Scenario: Create a resource with empty body
+    When I add "Content-Type" header equal to "application/ld+json"
+    And I send a "POST" request to "/dummies"
+    Then the response status code should be 400
+    And the JSON node "hydra:description" should be equal to 'Syntax error'
```
Comment (dunglas) [inline features/main/crud.feature]:
```suggestion
    And the JSON node "hydra:description" should be equal to "Syntax error"
```

File: features/main/crud.feature [hunk 5]
```diff
@@ -526,42 +532,8 @@ Feature: Create-Retrieve-Update-Delete
...
-      "relatedOwnedDummy": null,
-      "relatedOwningDummy": null,
-      "id": 1,
-      "name": "A nice dummy",
-      "alias": null,
-      "foo": null
-    }
-    """
+    Then the response status code should be 400
+    And the JSON node "hydra:description" should be equal to 'Syntax error'
```
Comment (dunglas) [inline features/main/crud.feature]:
```suggestion
    And the JSON node "hydra:description" should be equal to "Syntax error"
```

**Comment by soyuka** [comment]
This needs to be documented asap

**Comment by borNfreee** [comment]
> This needs to be documented asap

Here I'm crying again. Sorry - we are trying to be up-to-date with new stable releases in quite a big project.

Our tests failing for 2.4.3, and I spent quite a bit of time looking to the changelog, commits between [2.4.2 and 2.4.3](https://github.com/api-platform/core/compare/v2.4.2...v2.4.3) and searching in 3 repos something similar to this issue to find that this feature is not documented (I'm not even saying that this is a BC break in patch version).

Could you please revise the way you guys deploy new features? If it's released, *please* make sure it's documented. Please make sure there is a note in the changelog.

**Comment by dunglas** [comment]
@borNfreee I agree, I missed it wasn't targeting master. Maybe should we revert this patch and merge it only in master. Also, it introduces an inconsistency with GraphQL. WYDT @teohhanhui? Can I revert this patch safely?

**Comment by borNfreee** [comment]
If you guys decide to revert it, please tag a `2.4.4` if that is possible ASAP, since now everyone who wants to update a patch version will have a problem

**Comment by teohhanhui** [comment]
It's a bug fix. I don't see why it should be reverted.

But we should (I should) add the documentation.

**Comment by dunglas** [comment]
It breaks existing applications that are just upgrading to a new patch version, it’s not acceptable.

**Comment by teohhanhui** [comment]
It's only for custom operations which used an empty request body. Our previous hack introduced other bugs, such as throwing an unexpected and cryptic exception when the request body is empty.

There's no way to maintain compatibility with buggy behavior. So by definition a bug fix must break what depended on the buggy behavior.

**Comment by teohhanhui** [comment]
I know this is overused, but...



On a more serious note, fixing bugs inevitably breaks apps that were depending on the buggy behavior. We could try to minimize that impact, but it should not dictate how we should / should not fix a bug.

**Comment by dunglas** [comment]
We could at least agree to not merge bug fixes introducing BC breaks in patch releases. They should at least be in minor releases. This one is a true BC break, not just an edge case... because the old behavior was covered by a test.

It's too late to revert now, but could you add a note in the CHANGELOG explaining how to fix this BC break please?

**Comment by teohhanhui** [comment]
All bug fixes are BC breaks. We have many test cases that are wrong. It's an unavoidable problem.

> could you add a note in the CHANGELOG

I'm on it.

**Comment by teohhanhui** [comment]
See https://github.com/api-platform/core/pull/2815
